### PR TITLE
[cpu] Cpu device 1/n: memory allocation

### PR DESCRIPTION
**Pull request by AmesingFlank**
related issue = #2736 

This PR starts the work on `CpuDevice`. 

`allocate_memory` and `dealloc_memory` are implemented, and these are used in unified allocator.

The next PR will start moving the memory pool inside `CudeDevice` and `CpuDevice` ([discussion](https://github.com/taichi-dev/taichi/pull/2963#discussion_r713908020))

### CODE REVIEW

File: taichi/system/unified_allocator.cpp [hunk 1]
```diff
@@ -63,6 +64,18 @@ UnifiedAllocator::UnifiedAllocator(std::size_t size, Arch arch, Device *device)
 #else
     TI_NOT_IMPLEMENTED
 #endif
+  }
+  // This is an intermediate state.
+  // We will use memory pools to implement `Device::allocate_memory` soon.
+  else if (arch_ == Arch::x64 && false) {
```
Comment (ailzhang) [inline taichi/system/unified_allocator.cpp:70]: qq: when do we want to get rid of the `&& false` here?

**Review by ailzhang — APPROVED** [review]
Looks great, thanks!

**Comment by k-ye** [comment]
Let's associate these PRs with https://github.com/taichi-dev/taichi/issues/2736 :-)

File: taichi/system/unified_allocator.cpp (hunk 1 above)
Comment (AmesingFlank) [reply taichi/system/unified_allocator.cpp]: Ah nice catch. They shouldn't be here. I was debugging something and forgot to delete them.

**Review by k-ye — APPROVED** [review]
LGTM!

**Pull request merged**